src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -738,12 +913,174 @@ export class SavedObjectsRepository {
+    };
+
+    const updateResponse = await this._writeToCluster('update', {
+      id: rawId,
+      index: this.getIndexForType(type),
+      ...getExpectedVersionProperties(version, preflightResult),
+      refresh,
+      ignore: [404],
+      body: {
+        doc,
+      },
+    });
+
+    if (updateResponse.status === 404) {
+      // see "404s from missing index" above
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    return {};
+  }
+
+  async removeNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsRemoveNamespacesOptions = {}
+  ): Promise<{}> {
+    if (!this._allowedTypes.includes(type)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    if (!this._registry.isMultiNamespace(type)) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(`${type} doesn't support namespaces`);
+    }
+
+    if (!namespaces || !Array.isArray(namespaces) || !namespaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'namespaces must be a non-empty array of strings'
+      );
+    }
+
+    const { namespace, refresh = DEFAULT_REFRESH_SETTING } = options;
+
+    const rawId = this._serializer.generateRawId(undefined, type, id);
+    const preflightResult = await this.preflightCheckIncludesNamespace(type, id, namespace);
+    const existingNamespaces = getSavedObjectNamespaces(undefined, preflightResult);
+
+    if (existingNamespaces) {
+      // if there are somehow no existing namespaces, don't treat this as a missing namespace --
+      // instead, allow the operation to proceed and delete this saved object
+      const missing = namespaces.filter(x => !existingNamespaces.includes(x));
+      if (missing.length) {
+        throw SavedObjectsErrorHelpers.createBadRequestError(
+          `${rawId} doesn't exist in the following namespace(s): ${missing.join(', ')}`
+        );
+      }
+    }
+    const remainingNamespaces = existingNamespaces?.filter(x => !namespaces.includes(x));
+
+    if (remainingNamespaces?.length) {
+      // if there is 1 or more namespace remaining, update the saved object
+      const time = this._getCurrentTime();
+
+      const doc = {
+        updated_at: time,
+        namespaces: remainingNamespaces,
+      };
+
+      const updateResponse = await this._writeToCluster('update', {
+        id: rawId,
+        index: this.getIndexForType(type),
+        ...getExpectedVersionProperties(undefined, preflightResult),
+        refresh,
+        ignore: [404],
+        body: {
+          doc,
+        },
+      });
+
+      if (updateResponse.status === 404) {
+        // see "404s from missing index" above
+        throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+      }
+      return {};
+    } else {
+      // if there are no namespaces remaining, delete the saved object
+      const deleteResponse = await this._writeToCluster('delete', {
```

Comment:
I thought about it, but it makes testing logic a bit more complex -- then `addNamespaces` could result in either 2 operations (if updating a saved object) or 3 operations (if deleting a saved object), due to the fact that there would be an extra preflight `get` request. In that sense I think it's better to leave it as-is, rather than try to save a few more lines of code.